File: Ameekaa-App/app/src/main/java/com/ameekaa/app/DiarizationDynamicQuantize.kt
package com.ameekaa.app

import android.util.Log
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.*

data class DiarizationSegment(val startTime: Double, val endTime: Double)

class DynamicQuantizedDiarizationEngine(
    private val modelPath: String,
    private val config: Map<String, Any>,
    val audioProcessor: DynamicQuantizedAudioProcessor
) {
    private val TAG = "DynamicQuantizedDiarizationEngine"
    
    private val segmentLength: Double = (config["segment_length"] as? Double) ?: 2.0
    private val segmentStep: Double = (config["segment_step"] as? Double) ?: 2.0
    private val minSegmentRatio: Double = (config["min_segment_ratio"] as? Double) ?: 0.5
    
    val performanceStats = mutableMapOf<String, Any>(
        "total_inference_time" to 0.0,
        "total_segments_processed" to 0,
        "enrollment_time" to 0.0,
        "diarization_time" to 0.0
    )

    init {
        Log.i(TAG, "[SUCCESS] Initialized dynamic quantized diarization engine")
        Log.i(TAG, "[SUCCESS] Model performance: ${audioProcessor.benchmarkResults}")
    }

    fun computeSimilarity(embedding1: FloatArray, embedding2: FloatArray): Float {
        try {
            if (embedding1.isEmpty() || embedding2.isEmpty() || embedding1.size != embedding2.size) {
                Log.w(TAG, "Invalid embeddings for similarity computation")
                return 0.0f
            }
            
            val epsilon = 1e-8f
            
            // Calculate L2 norms (exactly like Python np.linalg.norm)
            var norm1 = 0.0f
            var norm2 = 0.0f
            for (i in embedding1.indices) {
                norm1 += embedding1[i] * embedding1[i]
                norm2 += embedding2[i] * embedding2[i]
            }
            norm1 = sqrt(norm1) + epsilon
            norm2 = sqrt(norm2) + epsilon

            // Normalize embeddings first (exactly like Python: embedding / norm)
            val emb1Norm = FloatArray(embedding1.size) { i -> embedding1[i] / norm1 }
            val emb2Norm = FloatArray(embedding2.size) { i -> embedding2[i] / norm2 }

            // Compute dot product of normalized embeddings (exactly like Python: np.dot)
            var similarity = 0.0f
            for (i in emb1Norm.indices) {
                similarity += emb1Norm[i] * emb2Norm[i]
            }

            return similarity.coerceIn(-1.0f, 1.0f)
        } catch (e: Exception) {
            Log.w(TAG, "Similarity computation failed: ${e.message}")
            return 0.0f
        }
    }

    fun diarizeMeeting(
        meetingPath: String,
        enrollmentEmbedding: FloatArray,
        speakerName: String,
        threshold: Float
    ): List<DiarizationSegment> {
        try {
            Log.i(TAG, "Diarizing meeting audio: $meetingPath")
            val startTime = System.currentTimeMillis()
            
            val (audio, sr) = audioProcessor.loadAudio(meetingPath)
            val duration = audio.size.toDouble() / sr
            Log.i(TAG, "Meeting duration: ${"%.2f".format(duration)}s")
            
            val segments = mutableListOf<DiarizationSegment>()
            var totalSegments = 0
            var matchedSegments = 0
            var totalInferenceTime = 0.0
            
            // Process segments using step size
            var currentTime = 0.0
            while (currentTime < duration) {
                val endTime = minOf(currentTime + segmentLength, duration)
                
                val startSample = (currentTime * sr).toInt()
                val endSample = (endTime * sr).toInt()
                val segmentAudio = audio.sliceArray(startSample until endSample)
                
                // Check minimum segment ratio
                if (segmentAudio.size < (segmentLength * sr * minSegmentRatio).toInt()) {
                    currentTime += segmentStep
                    continue
                }
                
                totalSegments++
                
                try {
                    val segmentStartTime = System.currentTimeMillis()
                    val segmentEmbedding = audioProcessor.extractEmbedding(segmentAudio, sr)
                    val segmentInferenceTime = System.currentTimeMillis() - segmentStartTime
                    totalInferenceTime += segmentInferenceTime
                    
                    val similarity = computeSimilarity(segmentEmbedding, enrollmentEmbedding)
                    Log.d(TAG, "Segment ${"%.2f".format(currentTime)}s-${"%.2f".format(endTime)}s | " +
                              "Similarity: ${"%.3f".format(similarity)} | " +
                              "Time: ${"%.1f".format(segmentInferenceTime.toDouble())}ms")
                    
                    if (similarity >= threshold) {
                        segments.add(DiarizationSegment(currentTime, endTime))
                        matchedSegments++
                        Log.i(TAG, "  -> Matched $speakerName (Similarity: ${"%.3f".format(similarity)})")
                    }
                    
                } catch (e: Exception) {
                    Log.w(TAG, "Failed to process segment ${"%.2f".format(currentTime)}s-${"%.2f".format(endTime)}s: ${e.message}")
                }
                
                currentTime += segmentStep
            }
            
            val diarizationTime = System.currentTimeMillis() - startTime
            performanceStats["diarization_time"] = diarizationTime.toDouble()
            performanceStats["total_inference_time"] = totalInferenceTime
            performanceStats["total_segments_processed"] = totalSegments
            
            val avgInferenceTime = if (totalSegments > 0) totalInferenceTime / totalSegments else 0.0
            val realTimeFactor = if (duration > 0) diarizationTime / (duration * 1000) else 0.0
            
            Log.i(TAG, "[SUCCESS] Diarization completed: $matchedSegments/$totalSegments segments matched")
            Log.i(TAG, "[SUCCESS] Processing time: ${"%.2f".format(diarizationTime / 1000.0)}s (RTF: ${"%.2f".format(realTimeFactor)}x)")
            Log.i(TAG, "[SUCCESS] Average inference time per segment: ${"%.1f".format(avgInferenceTime)}ms")
            
            return segments
            
        } catch (e: Exception) {
            throw DynamicQuantizedDiarizationError("Meeting diarization failed: ${e.message}")
        }
    }

    fun extractSegments(
        meetingPath: String,
        segments: List<DiarizationSegment>,
        outputPath: String
    ): Boolean {
        try {
            if (segments.isEmpty()) {
                Log.w(TAG, "No segments to extract")
                return false
            }
            
            val (audio, sr) = audioProcessor.loadAudio(meetingPath)
            val segmentAudioList = mutableListOf<FloatArray>()
            var totalDuration = 0.0
            
            for (segment in segments) {
                val startSample = (segment.startTime * sr).toInt()
                val endSample = (segment.endTime * sr).toInt()
                val segmentAudio = audio.sliceArray(startSample until endSample)
                segmentAudioList.add(segmentAudio)
                totalDuration += (segment.endTime - segment.startTime)
            }
            
            // Concatenate all segments
            val totalSamples = segmentAudioList.sumOf { it.size }
            val outputAudio = FloatArray(totalSamples)
            var currentIndex = 0
            for (segmentAudio in segmentAudioList) {
                segmentAudio.copyInto(outputAudio, currentIndex)
                currentIndex += segmentAudio.size
            }
            
            // Save as 16-bit PCM WAV file
            saveAudioAsWav(outputAudio, sr, outputPath)
            
            Log.i(TAG, "[SUCCESS] Extracted ${segments.size} segments (${"%.1f".format(totalDuration)}s total)")
            Log.i(TAG, "[SUCCESS] Saved to: $outputPath")
            return true
            
        } catch (e: Exception) {
            throw DynamicQuantizedDiarizationError("Failed to extract segments: ${e.message}")
        }
    }

    fun getPerformanceSummary(): Map<String, Any> {
        val totalSegments = performanceStats["total_segments_processed"] as? Int ?: 0
        val totalInferenceTime = performanceStats["total_inference_time"] as? Double ?: 0.0
        
        return mapOf(
            "enrollment_time" to (performanceStats["enrollment_time"] ?: 0.0),
            "diarization_time" to (performanceStats["diarization_time"] ?: 0.0),
            "total_inference_time" to totalInferenceTime,
            "total_segments_processed" to totalSegments,
            "avg_inference_time_per_segment" to if (totalSegments > 0) totalInferenceTime / totalSegments else 0.0,
            "model_benchmark" to audioProcessor.benchmarkResults
        )
    }

    private fun saveAudioAsWav(audioData: FloatArray, sampleRate: Int, outputPath: String) {
        try {
            val outputFile = File(outputPath)
            outputFile.parentFile?.mkdirs()
            
            // Create WAV header
            val dataSize = audioData.size * 2 // 16-bit samples
            val totalSize = 36 + dataSize
            
            FileOutputStream(outputFile).use { fos ->
                // Write WAV header
                fos.write("RIFF".toByteArray())
                fos.write(intToByteArray(totalSize))
                fos.write("WAVE".toByteArray())
                fos.write("fmt ".toByteArray())
                fos.write(intToByteArray(16)) // PCM header size
                fos.write(shortToByteArray(1)) // PCM format
                fos.write(shortToByteArray(1)) // Mono
                fos.write(intToByteArray(sampleRate))
                fos.write(intToByteArray(sampleRate * 2)) // Byte rate
                fos.write(shortToByteArray(2)) // Block align
                fos.write(shortToByteArray(16)) // Bits per sample
                fos.write("data".toByteArray())
                fos.write(intToByteArray(dataSize))
                
                // Write audio data as 16-bit PCM (match Python's sf.write behavior exactly)
                for (sample in audioData) {
                    // Convert float [-1, 1] to 16-bit PCM exactly like Python soundfile
                    // Ensure no clipping and preserve original audio quality
                    val normalizedSample = sample.coerceIn(-1.0f, 1.0f)
                    val pcmSample = (normalizedSample * 32767.0f).toInt().coerceIn(-32768, 32767).toShort()
                    fos.write(shortToByteArray(pcmSample))
                }
            }
        } catch (e: Exception) {
            throw DynamicQuantizedDiarizationError("Failed to save WAV file: ${e.message}")
        }
    }

    private fun intToByteArray(value: Int): ByteArray {
        return ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(value).array()
    }

    private fun shortToByteArray(value: Short): ByteArray {
        return ByteBuffer.allocate(2).order(ByteOrder.LITTLE_ENDIAN).putShort(value).array()
    }
} 

File: Ameekaa-Python/diarization_dynamic_quantize.py
import os
import numpy as np
import soundfile as sf
import logging
import time
from typing import List, Tuple, Dict, Any
from datetime import datetime

logger = logging.getLogger(__name__)

class DynamicQuantizedDiarizationError(Exception):
    pass

class DynamicQuantizedDiarizationEngine:
    def __init__(self, model_path: str, config: Dict[str, Any], audio_processor):
        self.config = config
        self.audio_processor = audio_processor
        self.segment_length = config['segment_length']
        self.segment_step = config['segment_step']
        self.min_segment_ratio = config['min_segment_ratio']
        self.performance_stats = {
            'total_inference_time': 0.0,
            'total_segments_processed': 0,
            'enrollment_time': 0.0,
            'diarization_time': 0.0
        }
        logger.info(f"[SUCCESS] Initialized dynamic quantized diarization engine")
        logger.info(f"[SUCCESS] Model performance: {getattr(self.audio_processor, 'benchmark_results', None)}")

    def compute_similarity(self, embedding1: np.ndarray, embedding2: np.ndarray) -> float:
        try:
            emb1_norm = embedding1 / (np.linalg.norm(embedding1) + 1e-8)
            emb2_norm = embedding2 / (np.linalg.norm(embedding2) + 1e-8)
            similarity = float(np.dot(emb1_norm, emb2_norm))
            return np.clip(similarity, -1.0, 1.0)
        except Exception as e:
            logger.warning(f"Similarity computation failed: {str(e)}")
            return 0.0

    def diarize_meeting(self, meeting_path: str, enrollment_embedding: np.ndarray, 
                       speaker_name: str, threshold: float) -> List[Tuple[float, float]]:
        try:
            logger.info(f"Diarizing meeting audio: {meeting_path}")
            start_time = time.time()
            audio, sr = self.audio_processor.load_audio(meeting_path)
            duration = len(audio) / sr
            logger.info(f"Meeting duration: {duration:.2f}s")
            segments = []
            total_segments = 0
            matched_segments = 0
            total_inference_time = 0.0
            for start in np.arange(0, duration, self.segment_step):
                end = min(start + self.segment_length, duration)
                segment_audio = audio[int(start*sr):int(end*sr)]
                if len(segment_audio) < int(self.segment_length * sr * self.min_segment_ratio):
                    continue
                total_segments += 1
                try:
                    segment_start_time = time.time()
                    segment_embedding = self.audio_processor.extract_embedding(segment_audio, sr)
                    segment_inference_time = time.time() - segment_start_time
                    total_inference_time += segment_inference_time
                    similarity = self.compute_similarity(segment_embedding, enrollment_embedding)
                    logger.debug(f"Segment {start:.2f}s-{end:.2f}s | Similarity: {similarity:.3f} | Time: {segment_inference_time*1000:.1f}ms")
                    if similarity >= threshold:
                        segments.append((start, end))
                        matched_segments += 1
                        logger.info(f"  -> Matched {speaker_name} (Similarity: {similarity:.3f})")
                except Exception as e:
                    logger.warning(f"Failed to process segment {start:.2f}s-{end:.2f}s: {str(e)}")
                    continue
            diarization_time = time.time() - start_time
            self.performance_stats['diarization_time'] = diarization_time
            self.performance_stats['total_inference_time'] = total_inference_time
            self.performance_stats['total_segments_processed'] = total_segments
            avg_inference_time = total_inference_time / total_segments if total_segments > 0 else 0
            real_time_factor = diarization_time / duration if duration > 0 else 0
            logger.info(f"[SUCCESS] Diarization completed: {matched_segments}/{total_segments} segments matched")
            logger.info(f"[SUCCESS] Processing time: {diarization_time:.2f}s (RTF: {real_time_factor:.2f}x)")
            logger.info(f"[SUCCESS] Average inference time per segment: {avg_inference_time*1000:.1f}ms")
            return segments
        except Exception as e:
            raise DynamicQuantizedDiarizationError(f"Meeting diarization failed: {str(e)}")

    def extract_segments(self, meeting_path: str, segments: List[Tuple[float, float]], 
                        output_path: str) -> bool:
        try:
            if not segments:
                logger.warning("No segments to extract")
                return False
            audio, sr = self.audio_processor.load_audio(meeting_path)
            segment_audio_list = []
            total_duration = 0
            for start, end in segments:
                start_sample = int(start * sr)
                end_sample = int(end * sr)
                segment_audio = audio[start_sample:end_sample]
                segment_audio_list.append(segment_audio)
                total_duration += (end - start)
            output_audio = np.concatenate(segment_audio_list)
            sf.write(output_path, output_audio, sr, subtype='PCM_16')
            logger.info(f"[SUCCESS] Extracted {len(segments)} segments ({total_duration:.1f}s total)")
            logger.info(f"[SUCCESS] Saved to: {output_path}")
            return True
        except Exception as e:
            raise DynamicQuantizedDiarizationError(f"Failed to extract segments: {str(e)}")

    def get_performance_summary(self) -> Dict[str, Any]:
        return {
            'enrollment_time': self.performance_stats['enrollment_time'],
            'diarization_time': self.performance_stats['diarization_time'],
            'total_inference_time': self.performance_stats['total_inference_time'],
            'total_segments_processed': self.performance_stats['total_segments_processed'],
            'avg_inference_time_per_segment': (
                self.performance_stats['total_inference_time'] / 
                self.performance_stats['total_segments_processed']
                if self.performance_stats['total_segments_processed'] > 0 else 0
            ),
            'model_benchmark': getattr(self.audio_processor, 'benchmark_results', None)
        } 
 